File: src/shared/models/card_cross_reference.py
# ============================================================================
# Source: COBOL copybook CVACT03Y.cpy — CARD-XREF-RECORD (RECLN 50)
# ============================================================================
# Mainframe-to-Cloud migration: VSAM KSDS → Aurora PostgreSQL
#
# Replaces the mainframe XREFFILE VSAM KSDS cluster (see app/jcl/XREFFILE.jcl)
# with a relational PostgreSQL table persisting the card → customer → account
# cross-reference. The alternate index XREFFILE.XREFAI02.PATH — which keyed
# cross-reference rows by their owning account ID in the VSAM world — is
# re-implemented here as a non-unique B-tree index
# (``ix_card_cross_reference_acct_id``) on the ``acct_id`` column.
# ============================================================================
# Copyright Amazon.com, Inc. or its affiliates.
# All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License").
# You may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================
"""SQLAlchemy 2.x ORM model for the ``card_cross_reference`` table.

Converts the COBOL copybook ``app/cpy/CVACT03Y.cpy`` (record layout
``CARD-XREF-RECORD``, 50-byte fixed-width record) to a SQLAlchemy 2.x
declarative ORM model representing a single card ↔ customer ↔ account
cross-reference row in the CardDemo Aurora PostgreSQL database.

The cross-reference table is the linking entity that resolves a physical
card number to its owning customer and account — a many-to-many bridge
that was implemented in the mainframe world as a dedicated VSAM KSDS
keyed by card number with an alternate index by account ID.

COBOL to Python Field Mapping
-----------------------------
==============  ==============  ==============  =========================
COBOL Field     COBOL Type      Python Column   SQLAlchemy Type
==============  ==============  ==============  =========================
XREF-CARD-NUM   ``PIC X(16)``   ``card_num``    ``String(16)`` — PK
XREF-CUST-ID    ``PIC 9(09)``   ``cust_id``     ``String(9)`` †
XREF-ACCT-ID    ``PIC 9(11)``   ``acct_id``     ``String(11)`` †
FILLER          ``PIC X(14)``   — (not mapped)  — (COBOL padding only)
==============  ==============  ==============  =========================

† **Numeric identifiers stored as strings.** The COBOL fields
  ``XREF-CUST-ID`` (``PIC 9(09)``) and ``XREF-ACCT-ID`` (``PIC 9(11)``)
  are numeric in the source record layout, but both are mapped to
  ``String(n)`` columns on the Python / PostgreSQL side. This is
  deliberate: migrated VSAM records can legitimately contain leading
  zeros (e.g., customer ID ``000000001`` or account ID
  ``00000000001``), and storing these as numeric types would silently
  strip the leading zeros at INSERT time, breaking the logical joins
  to :class:`~src.shared.models.customer.Customer` (``cust_id``) and
  :class:`~src.shared.models.account.Account` (``acct_id``).
  Byte-for-byte preservation of the original VSAM representation is
  required by AAP §0.7.1 ("preserve all existing functionality
  exactly as-is"). The identical convention is used for the companion
  FK-like identifiers on :class:`~src.shared.models.card.Card` and
  :class:`~src.shared.models.account.Account`.

Total RECLN: 16 + 9 + 11 + 14 = 50 bytes — matches the VSAM cluster
definition in ``app/jcl/XREFFILE.jcl`` (``RECSZ(50 50)``) and the
IDCAMS catalog entry in ``app/catlg/LISTCAT.txt``.

Alternate Index Preservation
----------------------------
The VSAM cluster ``XREFFILE`` defined an alternate index
``XREFFILE.XREFAI02.PATH`` keyed on ``XREF-ACCT-ID`` (one of 3 AIX
paths in the CardDemo mainframe layout — see the IDCAMS catalog
report ``app/catlg/LISTCAT.txt``). This alternate index enabled
efficient "find all cards mapped to account N" lookups from COBOL
programs:

* ``COACTVWC.cbl`` (F-004 Account View) — traverses the cross-reference
  by account ID to locate all cards on the account as part of the
  3-entity join (Account + Customer + Card) rendered on the account
  view screen.
* ``COCRDLIC.cbl`` (F-006 Card List) — similarly walks the
  cross-reference by account ID to produce the paginated 7-rows-per-page
  card listing scoped to the currently-selected account.
* ``COTRN02C.cbl`` (F-011 Transaction Add) — resolves the card-number
  input value to its owning account ID through the cross-reference at
  transaction creation time, so the posted transaction carries the
  correct ``acct_id`` for later balance updates and statement
  attribution.
* ``CBTRN02C.cbl`` / ``posttran_job.py`` (POSTTRAN batch stage) — the
  same card-number-to-account resolution is repeated in bulk during
  daily transaction posting; every inbound daily transaction must
  resolve to a valid cross-reference row or be rejected with reject
  code 102 (see the POSTTRAN 4-stage validation cascade).

To preserve this access pattern with Aurora PostgreSQL, this module
declares a non-unique B-tree index (``ix_card_cross_reference_acct_id``)
on the ``acct_id`` column via ``__table_args__``. The index is
intentionally non-unique: a single account can legitimately own
multiple cards (primary, authorised user, replacement, etc.), which
is exactly why the VSAM AIX was also non-unique (``NONUNIQUEKEY`` flag
in LISTCAT). The same index is separately created in the Flyway-style
migration ``db/migrations/V2__indexes.sql`` so that schema bootstrap
via ``Base.metadata.create_all()`` and migration-driven deployment
converge on identical relational shape.

Design Notes
------------
* Uses SQLAlchemy 2.x :func:`~sqlalchemy.orm.mapped_column` style with
  typed :class:`~sqlalchemy.orm.Mapped` annotations (NOT the legacy
  ``Column()`` constructor) — consistent with all sibling entity
  modules in ``src.shared.models``.
* ``Base`` is imported from the package ``__init__.py``
  (``src.shared.models``) so that this entity registers with the
  shared :class:`~sqlalchemy.MetaData` alongside the other CardDemo
  models, enabling unified ``Base.metadata.create_all()`` bootstrap
  and Alembic migration discovery.
* No ``FILLER`` column is mapped — the trailing 14 bytes of COBOL
  padding in the 50-byte VSAM record have no relational counterpart.
  In PostgreSQL, column widths are explicit and trailing padding
  carries no storage or semantic meaning.
* ``cust_id`` and ``acct_id`` are declared as plain ``String(n)``
  columns rather than as SQL-level ``ForeignKey`` references. The FK
  relationships to the :class:`~src.shared.models.customer.Customer`
  and :class:`~src.shared.models.account.Account` tables are real and
  enforced at the service layer, but the declarative FK is deliberately
  NOT wired at the ORM level here to keep this module and its
  referenced peers (``customer.py``, ``account.py``) independently
  loadable. This mirrors the pattern already established in
  ``card.py`` and ``account.py`` — referential integrity for the full
  set of relationships is managed by the Flyway-style migration
  scripts in ``db/migrations/``.
* This entity has no monetary fields — every column is a character
  identifier — so no ``decimal.Decimal`` / ``Numeric`` columns or
  financial-precision considerations apply.
* This entity has no optimistic-concurrency version column because
  cross-reference rows are write-once: created when a card is issued
  (during customer onboarding) and never updated in place. Deletion
  is the only terminal operation (on card closure), which is handled
  atomically by the service layer.
* Python 3.11+ only (aligned with AWS Glue 5.1 runtime and FastAPI
  deployment baseline).

See Also
--------
AAP §0.2.3 — Feature mapping for F-004 (Account View), F-006 (Card
List), F-011 (Transaction Add) — all consume this cross-reference.
AAP §0.5.1 — File-by-File Transformation Plan
(``card_cross_reference.py`` entry).
AAP §0.5.1 — DB migrations: ``db/migrations/V1__schema.sql`` (CREATE
TABLE), ``db/migrations/V2__indexes.sql`` (account-ID B-tree index),
``db/migrations/V3__seed_data.sql`` (seed from ``cardxref.txt``).
AAP §0.7.1 — Refactoring-Specific Rules (byte-for-byte fidelity to the
COBOL field widths; no enhancement beyond migration requirements).
``app/cpy/CVACT03Y.cpy`` — Original COBOL record layout (source
artifact, retained for traceability).
``app/jcl/XREFFILE.jcl`` — Original VSAM cluster definition
(``RECSZ(50 50)``, ``KEYS(16 0)``).
``app/catlg/LISTCAT.txt`` — IDCAMS catalog entry for
``XREFFILE.XREFAI02.PATH`` alternate index.
``app/data/ASCII/cardxref.txt`` — 50-row seed fixture loaded via
``db/migrations/V3__seed_data.sql``.
"""

from sqlalchemy import Index, String
from sqlalchemy.orm import Mapped, mapped_column

from src.shared.models import Base


class CardCrossReference(Base):
    """ORM entity for the ``card_cross_reference`` table (from COBOL ``CARD-XREF-RECORD``).

    Represents a single row in the CardDemo Aurora PostgreSQL
    ``card_cross_reference`` table, which replaces the mainframe VSAM KSDS
    ``XREFFILE`` dataset. Each row links exactly one credit card
    (``card_num``) to its owning customer (``cust_id``) and account
    (``acct_id``) — the many-to-many bridge that resolves card-level
    events to the customer and account contexts required by downstream
    business logic.

    Participates in the following online and batch flows:

    * **Online — F-004 Account View** (``COACTVWC`` →
      ``account_service.view()``): the cross-reference is traversed by
      ``acct_id`` (via the ``ix_card_cross_reference_acct_id`` index) to
      discover all cards on the account and assemble the consolidated
      3-entity view (Account + Customer + Card).
    * **Online — F-006 Card List** (``COCRDLIC`` →
      ``card_service.list_by_account()``): the cross-reference is
      traversed by ``acct_id`` to produce the paginated 7-rows-per-page
      card listing for the currently-selected account.
    * **Online — F-011 Transaction Add** (``COTRN02C`` →
      ``transaction_service.add()``): the cross-reference is queried by
      ``card_num`` (primary key) to resolve the user-supplied card
      number to its owning account ID before the new transaction is
      INSERTed — this is the mechanism that guarantees every
      transaction row carries a correct ``acct_id`` value.
    * **Batch — POSTTRAN** (``CBTRN02C`` → ``posttran_job.py``): the
      same card → account resolution is applied in bulk during daily
      transaction posting. Daily-transaction records whose card number
      does not resolve through the cross-reference are rejected with
      reject code 102 as part of the 4-stage validation cascade.

    Attributes
    ----------
    card_num : str
        **Primary key.** 16-character card number (from COBOL
        ``XREF-CARD-NUM``, ``PIC X(16)``). Stored as ``String(16)`` to
        preserve the original character representation byte-for-byte.
        Matches the VSAM cluster key length documented in
        ``app/jcl/XREFFILE.jcl`` (``KEYS(16 0)``) and the PK width of
        the companion :class:`~src.shared.models.card.Card` entity
        (``CVACT02Y.cpy``). There is exactly one cross-reference row
        per issued card — cross-reference rows are created at card
        issuance and deleted only on card closure.
    cust_id : str
        9-character zero-padded customer ID (from COBOL
        ``XREF-CUST-ID``, ``PIC 9(09)``). Stored as ``String(9)``
        rather than numeric to preserve leading zeros from migrated
        VSAM records. References
        :class:`~src.shared.models.customer.Customer.cust_id` at the
        service layer, though no database-level ``ForeignKey`` is
        declared here (see module docstring "Design Notes" — FK
        enforcement is managed by the Flyway migrations).
    acct_id : str
        11-character zero-padded owning account ID (from COBOL
        ``XREF-ACCT-ID``, ``PIC 9(11)``). Stored as ``String(11)``
        rather than numeric to preserve leading zeros from migrated
        VSAM records. References
        :class:`~src.shared.models.account.Account.acct_id` at the
        service layer. **Indexed** by
        ``ix_card_cross_reference_acct_id`` (non-unique B-tree) so that
        "find all cards mapped to account N" queries — issued by the
        F-004 Account View, F-006 Card List, F-011 Transaction Add,
        and POSTTRAN batch stage — remain efficient. This index
        directly replicates the VSAM alternate index
        ``XREFFILE.XREFAI02.PATH`` from the mainframe world (see the
        "Alternate Index Preservation" section of the module
        docstring).
    """

    __tablename__ = "card_cross_references"

    # ------------------------------------------------------------------
    # Non-unique B-tree index on the owning account ID.
    #
    # Replicates the VSAM alternate index XREFFILE.XREFAI02.PATH
    # (NONUNIQUEKEY) from the mainframe, enabling efficient
    # "find all cards mapped to account N" scans issued by:
    #
    #   * F-004 Account View   (COACTVWC → account_service.view())
    #   * F-006 Card List      (COCRDLIC → card_service.list_by_account())
    #   * F-011 Transaction Add (COTRN02C → transaction_service.add())
    #   * POSTTRAN batch stage (CBTRN02C → posttran_job.py)
    #
    # Declared non-unique because a single account can legitimately
    # own multiple cards (primary, authorised user, replacement card,
    # lost/stolen reissue, etc.) — mirroring the NONUNIQUEKEY property
    # of the original VSAM AIX (see app/catlg/LISTCAT.txt).
    # ------------------------------------------------------------------
    __table_args__ = (Index("ix_card_cross_reference_acct_id", "acct_id"),)

    # ------------------------------------------------------------------
    # Primary key: 16-character card number
    # (COBOL ``XREF-CARD-NUM`` PIC X(16))
    #
    # Stored as String(16) to preserve the original character
    # representation byte-for-byte. Matches the VSAM cluster key
    # length (KEYS(16 0)) in app/jcl/XREFFILE.jcl and the PK width
    # of the companion Card entity (CVACT02Y.cpy).
    # ------------------------------------------------------------------
    card_num: Mapped[str] = mapped_column(
        String(16),
        primary_key=True,
    )

    # ------------------------------------------------------------------
    # Customer ID (COBOL ``XREF-CUST-ID`` PIC 9(09))
    #
    # Stored as String(9) rather than numeric to preserve leading
    # zeros from migrated VSAM records (e.g., customer ID
    # '000000001'). References Customer.cust_id at the service layer;
    # no database-level FK is declared (see module docstring).
    # ------------------------------------------------------------------
    cust_id: Mapped[str] = mapped_column(
        String(9),
        nullable=False,
    )

    # ------------------------------------------------------------------
    # Owning account ID (COBOL ``XREF-ACCT-ID`` PIC 9(11))
    #
    # Stored as String(11) rather than numeric to preserve leading
    # zeros from migrated VSAM records (e.g., account ID
    # '00000000001'). References Account.acct_id at the service
    # layer. Indexed by ix_card_cross_reference_acct_id
    # (non-unique B-tree) above — directly replicates the VSAM
    # alternate index XREFFILE.XREFAI02.PATH.
    # ------------------------------------------------------------------
    acct_id: Mapped[str] = mapped_column(
        String(11),
        nullable=False,
    )

    # Note: COBOL ``FILLER PIC X(14)`` — the trailing 14 bytes of
    # padding in the original 50-byte VSAM record — is deliberately
    # NOT mapped. In the relational model, column widths are
    # explicit and trailing padding has no storage or semantic
    # meaning.

    def __repr__(self) -> str:  # pragma: no cover - trivial formatting
        """Return a developer-friendly string representation.

        Returns
        -------
        str
            Representation of the form
            ``CardCrossReference(card_num='4111111111111111',
            cust_id='000000001', acct_id='00000000001')``.
        """
        return f"CardCrossReference(card_num={self.card_num!r}, cust_id={self.cust_id!r}, acct_id={self.acct_id!r})"

File: carddemo-java/src/main/java/com/cardemo/model/entity/CardCrossReference.java
package com.cardemo.model.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import java.util.Objects;

/**
 * JPA entity mapping the legacy AWS CardDemo card cross-reference record onto the
 * PostgreSQL {@code card_cross_references} table.
 *
 * <p>This entity is the Java 25 / Spring Data JPA replacement for the VSAM KSDS
 * dataset {@code CARDXREF}, whose fixed 50-byte record layout is defined by the
 * COBOL copybook {@code app/cpy/CVACT03Y.cpy} ({@code 01 CARD-XREF-RECORD},
 * record length 50). It is the small but operationally critical <em>junction</em>
 * record that links a card to both its owning customer and its owning account. On
 * the mainframe this record was read by the online programs {@code COACTVWC}
 * (account view), {@code COCRDLIC} (card list / account-filtered browse) and
 * {@code COTRN02C} (add transaction), and by the batch programs {@code CBACT03C}
 * (cross-reference file read), {@code CBTRN02C} (daily transaction posting) and
 * {@code CBSTM03A} (statement generation).</p>
 *
 * <p>Two access paths existed in VSAM and both are preserved by the relational
 * mapping. The primary KSDS key is the 16-character card number
 * ({@code XREF-CARD-NUM}), which yields the card&nbsp;&rarr;&nbsp;(customer,
 * account) resolution and becomes this entity's sole {@code @Id}. The alternate
 * index {@code CXACAIX} is keyed on the account id ({@code XREF-ACCT-ID}) and
 * yields the reverse account&nbsp;&rarr;&nbsp;cards resolution used by the
 * account-filtered browses in {@code COACTVWC}/{@code COCRDLIC}/{@code COTRN02C}.
 * That alternate-index access path becomes the derived repository query
 * {@code CardCrossReferenceRepository.findByXrefAcctId(Long)} (a
 * {@code JpaRepository<CardCrossReference, String>}), backed by a secondary index
 * created on {@code account_id} in the Flyway {@code V2} migration (AAP §0.4.1).</p>
 *
 * <h2>Original COBOL layout (CVACT03Y.cpy &mdash; RECLN 50)</h2>
 * <pre>{@code
 * 01  CARD-XREF-RECORD.
 *     05  XREF-CARD-NUM            PIC X(16).
 *     05  XREF-CUST-ID             PIC 9(09).
 *     05  XREF-ACCT-ID             PIC 9(11).
 *     05  FILLER                   PIC X(14).
 * }</pre>
 *
 * <h2>Technology-substitution notes (Minimal Change Clause &mdash; AAP §0.7.1)</h2>
 * <ul>
 *   <li><strong>VSAM KSDS keyed access &rarr; JPA.</strong> The physically keyed
 *       {@code CARDXREF} cluster (16-byte key) becomes a relational table whose
 *       primary key is {@link #xrefCardNum}. Keyed reads served by CICS file
 *       control are replaced by a Spring Data {@code CardCrossReferenceRepository}
 *       ({@code JpaRepository<CardCrossReference, String>}).</li>
 *   <li><strong>Card number (PAN) &rarr; {@link String} primary key.</strong>
 *       {@code XREF-CARD-NUM PIC X(16)} is the 16-digit Primary Account Number. It
 *       is kept as a {@link String} (PostgreSQL {@code VARCHAR(16)}), not a numeric
 *       type, so leading characters are preserved exactly and the value matches the
 *       keyed VSAM access and {@code Card.cardNum}. It is the table's sole
 *       {@code @Id}.</li>
 *   <li><strong>Customer id &rarr; scalar {@link Long}, not an association.</strong>
 *       {@code XREF-CUST-ID PIC 9(09)} is an unsigned nine-digit identifier mapped
 *       to {@link Long} (PostgreSQL {@code BIGINT}), consistent with
 *       {@code Customer.custId}. It is modelled as a plain scalar column
 *       ({@code customer_id}) rather than a JPA {@code @ManyToOne} reference to
 *       {@code Customer}: this deliberately mirrors the flat COBOL/VSAM record (the
 *       cross-reference physically stored the customer id) and keeps this entity
 *       free of any compile-time coupling to {@code Customer}. The relationship to
 *       {@code customers} is a <em>logical</em> foreign key enforced in the DDL,
 *       not via an object reference.</li>
 *   <li><strong>Account id &rarr; scalar {@link Long}; the {@code CXACAIX}
 *       alternate-index column.</strong> {@code XREF-ACCT-ID PIC 9(11)} is an
 *       unsigned eleven-digit identifier mapped to {@link Long} (PostgreSQL
 *       {@code BIGINT}, because eleven digits exceed {@code Integer}'s
 *       ~2.1-billion ceiling) and column {@code account_id}, type-consistent with
 *       {@code Account.acctId}. This is the column on which the VSAM alternate
 *       index {@code CXACAIX} was keyed; it is the single most important contract
 *       on this entity. It is therefore the target of the derived query
 *       {@code findByXrefAcctId(Long)} and is backed by a dedicated secondary
 *       index in the Flyway {@code V2} migration so that account&rarr;cards lookups
 *       remain index-served. Like the customer id it is a <em>logical</em> foreign
 *       key (to {@code accounts.account_id}) and is deliberately
 *       <strong>not</strong> a {@code @ManyToOne} reference.</li>
 *   <li><strong>No optimistic locking.</strong> Per AAP §0.7.5, JPA
 *       {@code @Version} is applied <strong>only</strong> to {@code Account} and
 *       {@code Card} (the two read-update programs {@code COACTUPC} and
 *       {@code COCRDUPC}). The cross-reference record is never updated in place
 *       through a read-update snapshot comparison, so this entity intentionally
 *       carries <strong>no</strong> {@code @Version} column.</li>
 *   <li><strong>FILLER not materialized.</strong> The trailing
 *       {@code FILLER PIC X(14)} is reserved padding that pads the record to its
 *       50-byte length ({@code 16 + 9 + 11 + 14 = 50}); it carries no business
 *       data and is intentionally <strong>not</strong> mapped to a column. The
 *       50-byte record length is documented here for external-contract reference
 *       only (AAP §0.7.2).</li>
 * </ul>
 *
 * <h2>Primary-key &amp; column-name contract</h2>
 * <p>The {@link Column} names declared below &mdash; {@code card_number} (primary
 * key), {@code customer_id} and {@code account_id} &mdash; are authoritative for
 * the data layer. The Flyway {@code V1__create_schema.sql} cross-reference table
 * must declare {@code card_number} as the {@code VARCHAR(16)} primary key,
 * {@code customer_id} as {@code BIGINT} (logical foreign key to
 * {@code customers.customer_id}) and {@code account_id} as {@code BIGINT} (logical
 * foreign key to {@code accounts.account_id}); the {@code V2} migration adds a
 * secondary index on {@code account_id} to back the {@code CXACAIX} alternate-index
 * query; and the {@code V3} seed (loaded from {@code app/data/ASCII/cardxref.txt})
 * must align with these names and lengths.</p>
 *
 * <p>Per the Minimal Change Clause this entity is a pure persistence type: it
 * declares exactly the three mapped record fields, carries <strong>no</strong>
 * {@code @Version} column, carries no Jakarta Bean Validation annotations (input
 * validation lives in the request DTO layer, AAP §0.4.2), and models no JPA
 * associations &mdash; both foreign keys are plain scalar {@link Long} columns,
 * mirroring the original VSAM keyed-access pattern (no {@code @ManyToOne}).</p>
 *
 * <p><strong>Traceability.</strong> Derived from the frozen COBOL baseline at
 * commit SHA {@code 27d6c6f}. The COBOL source is read-only reference material and
 * is never copied into this repository.</p>
 *
 * @see Card
 * @see Account
 * @see Customer
 */
@Entity
@Table(name = "card_xref")
public class CardCrossReference {

    /**
     * Primary key &mdash; the 16-digit card number (Primary Account Number) that
     * links this cross-reference to a single card.
     *
     * <p>Migrated from {@code XREF-CARD-NUM PIC X(16)}. Kept as a {@link String}
     * (PostgreSQL {@code VARCHAR(16)}) rather than a numeric type so that any
     * leading characters are preserved exactly and the value matches both the keyed
     * VSAM access and {@code Card.cardNum}. The downstream repository is therefore
     * {@code JpaRepository<CardCrossReference, String>}.</p>
     */
    // XREF-CARD-NUM PIC X(16) -> 16-char card number (PAN), primary VSAM key -> String(16) (VARCHAR(16)) @Id
    @Id
    @Column(name = "card_number", length = 16, nullable = false)
    private String xrefCardNum;

    /**
     * Owning customer identifier &mdash; a scalar foreign key, not an association.
     *
     * <p>Migrated from {@code XREF-CUST-ID PIC 9(09)}: an unsigned nine-digit
     * identifier, mapped to {@link Long} (PostgreSQL {@code BIGINT}) for
     * consistency with {@code Customer.custId}. This is a <em>logical</em> foreign
     * key to {@code customers.customer_id} (scalar, per the flat VSAM contract); it
     * is deliberately <strong>not</strong> a {@code @ManyToOne} reference, so this
     * entity imports nothing from {@code Customer}.</p>
     */
    // XREF-CUST-ID PIC 9(09) -> 9-digit owning customer id -> Long (BIGINT); scalar FK to customers.customer_id (NOT @ManyToOne)
    @Column(name = "customer_id", nullable = false)
    private Long xrefCustId;

    /**
     * Owning account identifier &mdash; a scalar foreign key and the
     * {@code CXACAIX} alternate-index column.
     *
     * <p>Migrated from {@code XREF-ACCT-ID PIC 9(11)}: an unsigned eleven-digit
     * identifier, mapped to {@link Long} (PostgreSQL {@code BIGINT}) because eleven
     * digits exceed {@code Integer}'s ~2.1-billion ceiling and to stay
     * type-consistent with {@code Account.acctId}. This is a <em>logical</em>
     * foreign key to {@code accounts.account_id} and is deliberately
     * <strong>not</strong> a {@code @ManyToOne} reference.</p>
     *
     * <p><strong>{@code XREF-ACCT-ID} = {@code CXACAIX} alternate index &rarr;
     * {@code findByXrefAcctId}.</strong> On the mainframe the {@code CARDXREF}
     * cluster carried the alternate index {@code CXACAIX} keyed on this field,
     * enabling the account&rarr;cards lookup used by {@code COACTVWC},
     * {@code COCRDLIC} and {@code COTRN02C}. The
     * {@code CardCrossReferenceRepository} therefore exposes a derived query
     * {@code findByXrefAcctId(Long)} (and/or {@code findAllByXrefAcctId}) over this
     * {@code account_id} column, backed by a dedicated secondary index created in
     * the Flyway {@code V2} migration so the reverse lookup remains index-served.</p>
     */
    // XREF-ACCT-ID PIC 9(11) -> 11-digit owning account id -> Long (BIGINT); scalar FK to accounts.account_id; CXACAIX alternate-index column -> findByXrefAcctId (secondary index in Flyway V2)
    @Column(name = "account_id", nullable = false)
    private Long xrefAcctId;

    /**
     * Default no-argument constructor required by the JPA provider (Hibernate) to
     * instantiate the entity reflectively before populating its fields.
     */
    public CardCrossReference() {
        // Intentionally empty: JPA/Hibernate instantiates then sets fields.
    }

    /**
     * Returns the primary-key card number ({@code XREF-CARD-NUM}).
     *
     * @return the 16-character card number, or {@code null} if unset
     */
    public String getXrefCardNum() {
        return xrefCardNum;
    }

    /**
     * Sets the primary-key card number ({@code XREF-CARD-NUM}).
     *
     * @param xrefCardNum the 16-character card number to set
     */
    public void setXrefCardNum(String xrefCardNum) {
        this.xrefCardNum = xrefCardNum;
    }

    /**
     * Returns the owning customer identifier ({@code XREF-CUST-ID}).
     *
     * <p>This is a scalar foreign key to {@code customers.customer_id}; it is not a
     * JPA association.</p>
     *
     * @return the owning customer identifier, or {@code null} if unset
     */
    public Long getXrefCustId() {
        return xrefCustId;
    }

    /**
     * Sets the owning customer identifier ({@code XREF-CUST-ID}).
     *
     * @param xrefCustId the owning customer identifier to set
     */
    public void setXrefCustId(Long xrefCustId) {
        this.xrefCustId = xrefCustId;
    }

    /**
     * Returns the owning account identifier ({@code XREF-ACCT-ID}).
     *
     * <p>This is a scalar foreign key to {@code accounts.account_id} and the
     * {@code CXACAIX} alternate-index column; it is not a JPA association. The
     * repository's {@code findByXrefAcctId(Long)} query selects on this value.</p>
     *
     * @return the owning account identifier, or {@code null} if unset
     */
    public Long getXrefAcctId() {
        return xrefAcctId;
    }

    /**
     * Sets the owning account identifier ({@code XREF-ACCT-ID}).
     *
     * @param xrefAcctId the owning account identifier to set
     */
    public void setXrefAcctId(Long xrefAcctId) {
        this.xrefAcctId = xrefAcctId;
    }

    /**
     * Identity-based equality keyed on the card number ({@link #xrefCardNum}).
     *
     * <p>Two {@code CardCrossReference} instances are equal when they are of the
     * exact same class and share the same {@link #xrefCardNum}. The primary key
     * alone defines entity identity; the mutable id columns are deliberately
     * excluded so equality stays stable. Exact-class comparison (rather than
     * {@code instanceof}) is used so a proxy/subclass is not treated as equal to a
     * different concrete type.</p>
     *
     * @param o the object to compare with
     * @return {@code true} if {@code o} is a {@code CardCrossReference} with an
     *         equal card number
     */
    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        CardCrossReference that = (CardCrossReference) o;
        return Objects.equals(xrefCardNum, that.xrefCardNum);
    }

    /**
     * Hash code derived solely from {@link #xrefCardNum}, consistent with
     * {@link #equals(Object)}.
     *
     * @return the hash code of the card number
     */
    @Override
    public int hashCode() {
        return Objects.hash(xrefCardNum);
    }

    /**
     * Diagnostic representation that intentionally protects cardholder data. The
     * card number (PAN) is masked to its last four digits via
     * {@link #maskedCardNumber()}; only the masked PAN and the two non-sensitive
     * owning identifiers are included, so this value is safe to emit to logs.
     *
     * @return a human-readable, non-sensitive description of this cross-reference
     */
    @Override
    public String toString() {
        return "CardCrossReference{"
                + "xrefCardNum='" + maskedCardNumber() + '\''
                + ", xrefCustId=" + xrefCustId
                + ", xrefAcctId=" + xrefAcctId
                + '}';
    }

    /**
     * Masks the card number (PAN) for safe inclusion in {@link #toString()},
     * revealing at most the final four digits and replacing every preceding
     * character with {@code '*'}. A {@code null} value yields {@code "null"}, and a
     * value of four or fewer characters is fully masked so no digits leak.
     *
     * @return the masked card number suitable for logging
     */
    private String maskedCardNumber() {
        if (xrefCardNum == null) {
            return "null";
        }
        int length = xrefCardNum.length();
        if (length <= 4) {
            // Too short to reveal a last-four suffix safely; mask completely.
            return "*".repeat(length);
        }
        return "*".repeat(length - 4) + xrefCardNum.substring(length - 4);
    }
}
